Sources/SwiftInspectorVisitors/TypeDescription.swift

```diff
@@ -0,0 +1,267 @@
+// Created by Dan Federman on 2/2/21.
+//
+// Copyright © 2021 Dan Federman
+//
+// Distributed under the MIT License
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+import SwiftSyntax
+
+/// An enum that describes a parsed type in a cannonical form.
+public enum TypeDescription: Codable, Equatable, CustomStringConvertible {
+  /// A root type. e.g. Int
+  case simple(name: String)
+  /// A nested type. e.g. Array.Element
+  indirect case nested(name: String, parentType: TypeDescription)
+  /// A composed type. e.g. Identifiable & Equatable
+  indirect case composition([TypeDescription])
+  /// An optional type. e.g. Int?
+  indirect case optional(TypeDescription)
+  /// An implicitly unwrapped optional type. eg. Int!
+  indirect case implicitlyUnwrappedOptional(TypeDescription)
+  /// An array. e.g. [Int]
+  indirect case array(TypeDescription)
+  /// A dictionary. e.g. [Int: String]
+  indirect case dictionary(key: TypeDescription, value: TypeDescription)
+  /// A tuple. e.g. (Int, String)
+  indirect case tuple([TypeDescription])
+  /// A type that can't be represented by the above cases.
+  case unknown(text: String)
+
+  /*
+   * Note that we do not yet support the following syntax types:
+   * SomeTypeSyntax
+   * MetatypeTypeSyntax
+   * FunctionTypeSyntax
+   * AttributedTypeSyntax
+   * UnknownTypeSyntax
+   *
+   * We will likely need to add these types at some point in the future.
+   * We’ll Get There™
+   */
+
+  /// A description of this type that can be used for code-generation to represent this type.
```

Comment:
with the latest update I'm less sure that's required. lmk what you think.

Reply:
@dfed I think it'd be nice to have this documentation. One of the main things about SwiftSyntax is that whatever you _read_ you can _write_ and it will output the same thing always see [this](https://github.com/apple/swift/blob/main/lib/Syntax/README.md#design-and-implementation-guidelines)

> Represent Swift source with "full fidelity" - parsing a source file and printing the syntax tree should result in the same file.

We're deviating from this by not representing the exact notation of the source code which is totally fine for our use case but we're also being opinionated on style here which is something we should at least put in a comment IMHO

Reply:
Particularly in this https://github.com/fdiaz/SwiftInspector/pull/69/files#diff-6a8854acd551d0fde3152864ddc0e807bae90c7b740b33078bffb630c1798356R76

Reply:
Okay. Let me better document this in https://github.com/fdiaz/SwiftInspector/pull/71 since then I can say that the output is a canonical representation (which indicates that it isn't necessarily the same as the original)